src/core/meshcop/joiner.cpp

```diff
@@ -234,31 +236,49 @@ void Joiner::SendJoinerFinalize(void)
     stateTlv->SetState(MeshCoP::StateTlv::kAccept);
     cur += sizeof(*stateTlv);
 
+    length = mNetif.GetDtls().mProvisioningUrl.GetLength();
+
+    if (length > 0)
+    {
+        length += sizeof(Tlv);
+        memcpy(cur, &mNetif.GetDtls().mProvisioningUrl, length);
+        cur += length;
+    }
+
     mNetif.GetDtls().Send(buf, static_cast<uint16_t>(cur - buf));
 
     otLogInfoMeshCoP("Sent joiner finalize\r\n");
 }
 
 void Joiner::ReceiveJoinerFinalizeResponse(uint8_t *buf, uint16_t length)
 {
-    Message *message;
+    Message *message = NULL;
     Coap::Header header;
+    StateTlv state;
 
     VerifyOrExit((message = mNetif.GetIp6().mMessagePool.New(Message::kTypeIp6, 0)) != NULL, ;);
     SuccessOrExit(message->Append(buf, length));
     SuccessOrExit(header.FromMessage(*message));
+    SuccessOrExit(message->SetOffset(header.GetLength()));
 
     VerifyOrExit(header.GetType() == Coap::Header::kTypeAcknowledgment &&
                  header.GetCode() == Coap::Header::kCodeChanged &&
                  header.GetMessageId() == 0 &&
                  header.GetTokenLength() == 0, ;);
 
-    otLogInfoMeshCoP("received joiner finalize response\r\n");
+    SuccessOrExit(Tlv::GetTlv(*message, Tlv::kState, sizeof(state), state));
+    VerifyOrExit(state.IsValid(), ;);
```

Comment:
We may have a TODO here: to check the state, and go on with scan process if the state is Reject.